aggregator/src/aggregator/report_writer.rs

```diff
@@ -114,32 +120,53 @@ impl<C: Clock> ReportWriteBatcher<C> {
     #[tracing::instrument(name = "ReportWriteBatcher::write_batch", skip_all)]
     async fn write_batch(
         ds: Arc<Datastore<C>>,
-        report_writers: Vec<Box<dyn ReportWriter<C>>>,
-        result_txs: Vec<oneshot::Sender<Result<(), Arc<Error>>>>,
+        counter_shard_count: u64,
+        mut report_results: Vec<(ReportResult<C>, ResultSender)>,
     ) {
-        // Check preconditions.
-        assert_eq!(report_writers.len(), result_txs.len());
+        let ord = thread_rng().gen_range(0..counter_shard_count);
+
+        // Sort by task ID to prevent deadlocks with concurrently running transactions. Since we are
```

Comment:
Idle thought: we may eventually want to use a B-Tree (i.e. `BTreeSet`) ordering on task ID, rather than a `Vec`, to amortize the cost of sorting across the uploads. I think batches are likely to stay small enough that sorting is "cheap", though.

Reply:
I did think about this (although with a less optimal BinaryHeap) but yeah I figured the batches are small enough where the additional overhead of a complex data structure would prove regressive at small batch size. Would likely need to dive into a profiler to know one way or the other. I'm planning to do this anyways for other reasons, will check this out along the way and open a PR if it can be improved.